src/parser/parser.ee.ts

```diff
@@ -98,3 +98,225 @@ export async function parseStream({
     }
   }
 }
+
+/**
+ * Filter files in the zip based on a list of substrings
+ * @param zipFiles List of files in the zip
+ * @param filterList List of substrings to filter the files
+ * @returns The matched file or undefined
+ */
+function filterFiles(
+  zipFiles: { [key: string]: Uint8Array },
+  filterList: string[],
+): Uint8Array | undefined {
+  for (const fileName of Object.keys(zipFiles)) {
+    if (
+      filterList.some((filter) => {
+        return fileName.includes(filter);
+      })
+    ) {
+      return zipFiles[fileName];
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Parse a .docx file and extract text content
+ * @param buffer - The binary content of the .docx file
+ * @returns Extracted text as a string
+ */
+async function parseDocx(buffer: Buffer): Promise<string> {
+  const zipFiles: Unzipped = await new Promise((resolve, reject) => {
+    unzip(new Uint8Array(buffer), (error, files) => {
+      if (error) {
+        reject(error);
+      } else {
+        resolve(files);
+      }
+    });
+  });
+  const mainDocument = filterFiles(zipFiles, ['word/document']);
+
+  if (!mainDocument) {
+    throw new Error('Invalid .docx file: No document file found');
+  }
+
+  const parser = new XMLParser({
+    attributeNamePrefix: '@_',
+    ignoreAttributes: false,
+    removeNSPrefix: true,
+  });
+  const jsonContent = parser.parse(
+    new TextDecoder('utf8').decode(mainDocument),
+  ) as XMLNode;
+
+  const extractedText = extractTextFromJson(jsonContent);
+  return formatText(extractedText);
+}
+
+/**
+ * Extract plain text from the parsed XML JSON object
+ * @param node - The parsed JSON object from the XML content
+ * @returns Extracted plain text with preserved newlines
+ */
+function extractTextFromJson(node: XMLNode): string {
+  let text = '';
+
+  /**
+   * Recursive function to traverse nodes and extract text
+   * @param node The current XML node
+   */
+  function traverse(node: XMLNode): void {
+    if (Array.isArray(node)) {
+      for (const child of node) {
+        traverse(child);
+      }
+    } else if (typeof node === 'object') {
+      const body = node.body;
+
+      if (body && typeof body === 'object' && !Array.isArray(body)) {
+        if ('p' in body) {
+          const paragraphs = body.p;
+          text += Array.isArray(paragraphs)
+            ? parseParagraphs(paragraphs)
+            : extractParagraphText(paragraphs) + '\n';
+        }
+      } else if ('p' in node) {
```

Comment:
Here you are dealing with only `p` but just like HTML, the office formats have more tags available, making this version of your code more complicated.

you need to adapt your whole function (not just this part) to be able to add to the text the content of the node when the node is `p` or `h`, or another when we found new relevant nodes in the future

Reply:
additionally, please copy the code from the ansearch repo. I took a long time to make sure the code there is readable, and I can see from your code that you understand it and you are not just mindlessly copy pasting.

Just be careful when copying it, since it uses async maps, which are highly discouraged for unknown objects (I learned that after doing the code for ansearch)

Make sure you also remove the `fixSentenceReturn` from my code, or update it to add dots only when there is a line return (not at the end of every sentences)